### PR TITLE
Candle indicators

### PR DESCRIPTION
**Pull request by btezergil**
Changes proposed in this pull request:
- Added three inside up/down candlestick indicators
- Added morning/evening star candlestick indicators
- Added bullish/bearish kicker candlestick indicators
- Added piercing/dark cloud candlestick indicators

- [x] added an entry with related ticket number(s) to the unreleased section of `CHANGES.md`

### CODE REVIEW

File: ta4j-core/src/main/java/org/ta4j/core/indicators/candles/BearishKickerIndicator.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    public BearishKickerIndicator(final BarSeries series) {
+        super(series);
+        this.trendIndicator = new UpTrendIndicator(series);
+        this.realBodyIndicator = new RealBodyIndicator(series);
+        this.bigBodyThresholdPercentage = series.numFactory().numOf(0.03);
+    }
+
+    @Override
+    protected Boolean calculate(int index) {
+        if (index < 1) {
```
Comment (TheCookieLab) [inline ta4j-core/src/main/java/org/ta4j/core/indicators/candles/BearishKickerIndicator.java]: Let's move this threshold to the getCountOfUnstableBars(). See RSIIndicator for an example. This comment applies universally to all new classes in this PR.

File: ta4j-examples/.factorypath [hunk 2]
```diff
@@ -0,0 +1,17 @@
+<factorypath>
```
Comment (TheCookieLab) [inline ta4j-examples/.factorypath]: .gitignore this

File: ta4j-core/src/main/java/org/ta4j/core/indicators/candles/BearishKickerIndicator.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    /**
+     * Constructor.
+     *
+     * @param series the bar series
+     */
+    public BearishKickerIndicator(final BarSeries series) {
+        super(series);
+        this.trendIndicator = new UpTrendIndicator(series);
+        this.realBodyIndicator = new RealBodyIndicator(series);
+        this.bigBodyThresholdPercentage = series.numFactory().numOf(0.03);
```
Comment (TheCookieLab) [inline ta4j-core/src/main/java/org/ta4j/core/indicators/candles/BearishKickerIndicator.java]: Rather than hardcode this to 0.03, let's make it configurable via optional constructor param. If not provided we can use 0.03 as the default. This comment also applies to the other classes.

File: ta4j-core/src/main/java/org/ta4j/core/indicators/candles/MorningStarIndicator.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        Bar secondBar = getBarSeries().getBar(index - 1);
+        Bar thirdBar = getBarSeries().getBar(index);
+        Num firstBarPercentage = this.realBodyIndicator.getValue(index - 2).abs().dividedBy(firstBar.getOpenPrice());
+        Num secondBarPercentage = this.realBodyIndicator.getValue(index - 1).abs().dividedBy(secondBar.getOpenPrice());
+        Num thirdBarPercentage = this.realBodyIndicator.getValue(index).abs().dividedBy(thirdBar.getOpenPrice());
+
+        return firstBar.isBearish() && firstBarPercentage.isGreaterThanOrEqual(bigBodyThresholdPercentage)
+                && secondBar.getOpenPrice().isLessThan(firstBar.getClosePrice())
+                && secondBarPercentage.isLessThanOrEqual(smallBodyThresholdPercentage)
+                && thirdBar.getClosePrice().isLessThan(firstBar.getOpenPrice()) && thirdBar.isBullish()
```
Comment (TheCookieLab) [inline ta4j-core/src/main/java/org/ta4j/core/indicators/candles/MorningStarIndicator.java]: I'm certainly no expert on candle indicators but for a bullish reversal should the 3rd bar's close price be less than the 1st bar's open? It seems it should be the other way around, i.e. the 3rd bar's close should be higher than the 1st bar's open.

File: ta4j-core/src/main/java/org/ta4j/core/indicators/candles/EveningStarIndicator.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        Bar secondBar = getBarSeries().getBar(index - 1);
+        Bar thirdBar = getBarSeries().getBar(index);
+        Num firstBarPercentage = this.realBodyIndicator.getValue(index - 2).abs().dividedBy(firstBar.getOpenPrice());
+        Num secondBarPercentage = this.realBodyIndicator.getValue(index - 1).abs().dividedBy(secondBar.getOpenPrice());
+        Num thirdBarPercentage = this.realBodyIndicator.getValue(index).abs().dividedBy(thirdBar.getOpenPrice());
+
+        return firstBar.isBullish() && firstBarPercentage.isGreaterThanOrEqual(bigBodyThresholdPercentage)
+                && secondBar.getOpenPrice().isGreaterThan(firstBar.getClosePrice())
+                && secondBarPercentage.isLessThanOrEqual(smallBodyThresholdPercentage)
+                && thirdBar.getClosePrice().isGreaterThan(firstBar.getOpenPrice()) && thirdBar.isBearish()
```
Comment (TheCookieLab) [inline ta4j-core/src/main/java/org/ta4j/core/indicators/candles/EveningStarIndicator.java]: Similar to the comment below for the MorningStarIndicator, during a bearish reversal shouldn't the 3rd bar close comparable or lower than the 1st bar? What you use specifically as the reference point - open vs close (or even midpoint) can be up for debate.

File: ta4j-core/src/test/java/org/ta4j/core/indicators/candles/EveningStarIndicatorTest.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        List<Bar> bars = new ArrayList<Bar>(30);
+        for (int i = 0; i < 17; ++i) {
+            bars.add(
+                    new MockBarBuilder(numFactory).openPrice(i).closePrice(i + 6).highPrice(i + 8).lowPrice(i).build());
+        }
+
+        return bars;
+    }
+
+    @Test
```
Comment (TheCookieLab) [inline ta4j-core/src/test/java/org/ta4j/core/indicators/candles/EveningStarIndicatorTest.java:66]: please add more tests for negative and edge-case scenarios. Unit tests not only demonstrate the code is doing what it's intended to, but just as importantly serves as a shield against regressions during future changes.

File: ta4j-core/src/main/java/org/ta4j/core/indicators/candles/ThreeInsideDownIndicator.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        this.trendIndicator = new UpTrendIndicator(series);
+    }
+
+    @Override
+    protected Boolean calculate(int index) {
+        if (index < 2) {
+            // Three inside down is a 3-candle pattern
+            return false;
+        }
+        BearishHaramiIndicator harami = new BearishHaramiIndicator(getBarSeries());
```
Comment (TheCookieLab) [inline ta4j-core/src/main/java/org/ta4j/core/indicators/candles/ThreeInsideDownIndicator.java]: Let's make this a class property (final) rather than instantiate every time in the hot path.

File: ta4j-core/src/main/java/org/ta4j/core/indicators/candles/ThreeInsideUpIndicator.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        this.trendIndicator = new DownTrendIndicator(series);
+    }
+
+    @Override
+    protected Boolean calculate(int index) {
+        if (index < 2) {
+            // Three inside up is a 3-candle pattern
+            return false;
+        }
+        BullishHaramiIndicator harami = new BullishHaramiIndicator(getBarSeries());
```
Comment (TheCookieLab) [inline ta4j-core/src/main/java/org/ta4j/core/indicators/candles/ThreeInsideUpIndicator.java]: same as above

File: ta4j-core/src/main/java/org/ta4j/core/indicators/candles/PiercingIndicator.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+ * Piercing candle indicator.
+ *
+ * @see <a href="https://www.investopedia.com/terms/piercing-pattern.asp">
+ *      https://www.investopedia.com/terms/piercing-pattern.asp</a>
+ */
+public class PiercingIndicator extends CachedIndicator<Boolean> {
+
+    private final DownTrendIndicator trendIndicator;
+    private final RealBodyIndicator realBodyIndicator;
+    private Num bigBodyThresholdPercentage;
```
Comment (TheCookieLab) [inline ta4j-core/src/main/java/org/ta4j/core/indicators/candles/PiercingIndicator.java]: Can be final.

File: ta4j-core/src/main/java/org/ta4j/core/indicators/candles/BearishKickerIndicator.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+import org.ta4j.core.indicators.CachedIndicator;
+import org.ta4j.core.indicators.trend.UpTrendIndicator;
+import org.ta4j.core.num.Num;
+
+/**
+ * Bearish kicker candle indicator.
+ *
+ * @see <a href="https://www.investopedia.com/terms/k/kickerpattern.asp">
+ *      https://www.investopedia.com/terms/k/kickerpattern.asp</a>
+ */
```
Comment (TheCookieLab) [inline ta4j-core/src/main/java/org/ta4j/core/indicators/candles/BearishKickerIndicator.java:38]: Please tag these new classes with @since x.xx

File: ta4j-core/src/main/java/org/ta4j/core/indicators/candles/MorningStarIndicator.java (hunk 4 above)
Comment (btezergil) [reply ta4j-core/src/main/java/org/ta4j/core/indicators/candles/MorningStarIndicator.java]: I've made some research before implementing this, and checked some public sites again. While multiple say that it should be higher than the middle point, others also add that it should be less than the first candle's open. I think the close point is not the main focus here. All content basically agree on the candle needing to be bullish + having a big body. 

I think the best course of action is to remove that requirement altogether, or change it so that it's higher than the middle point. The closing point of the third candle seems to be up for debate. We can discuss this further, I'm open to change this.

File: ta4j-core/src/test/java/org/ta4j/core/indicators/candles/EveningStarIndicatorTest.java (hunk 6 above)
Comment (btezergil) [reply ta4j-core/src/test/java/org/ta4j/core/indicators/candles/EveningStarIndicatorTest.java:66]: I've pushed all other review items except this one. I will add some extra tests - do you need the tests only for morning/evening star or for all classes?

**Comment by TheCookieLab** [comment]
All classes.
